File: src/Engine/Managers/TaskManager/TaskManager.java
package Engine.Managers.TaskManager;

import Engine.Managers.Manager;

import java.util.ArrayList;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

/**
 * Created by steppers on 01/11/15.
 */
public class TaskManager extends Manager {

    private static TaskManager instance = new TaskManager();
    public static TaskManager getInstance(){return instance;}

    private TaskThread[] threads;
    private BlockingQueue<Task> tasks;

    public void init()
    {
        threads = new TaskThread[Runtime.getRuntime().availableProcessors()];
        tasks = new LinkedBlockingQueue<>();
        for(int i = 0; i < threads.length; i++)
        {
            threads[i] = new TaskThread(i);
            threads[i].start();
        }
    }

    public void waitForTaskCompletion()
    {
        while(!tasks.isEmpty()){}
        for(TaskThread thread : threads)
        {
            while(true)
            {
                synchronized (thread.hasTask)
                {
                    if(!thread.hasTask)
                        break;
                }
            }
        }
    }

    public void submitTasks(ArrayList<Task> tasks)
    {
        for(Task t : tasks)
        {
            this.tasks.add(t);
        }
    }

    public void stop()
    {
        while(!tasks.isEmpty()){}
        for(Thread t : threads)
        {
            t.interrupt();
        }
    }

    public class TaskThread extends Thread
    {
        private int id;
        private Task t;
        public Boolean hasTask = false;

        public TaskThread(int id)
        {
            this.id = id;
        }

        public void run()
        {
            while(true)
            {
                try {
                    synchronized (hasTask){hasTask = false;}
                    t = tasks.take();
                    synchronized (hasTask){hasTask = true;}
                    t.execute(this);
                } catch (InterruptedException e) {
                    break;
                }
            }
        }

        public int getThreadId()
        {
            return id;
        }
    }
}

File: src/Engine/Managers/TaskManager/TaskManager.cpp
//
// Created by steppers on 05/11/15.
//

#include "TaskManager.h"
#include "TerminateTask.h"

TaskManager::TaskManager() { }
TaskManager::~TaskManager() { }

void TaskManager::init() {
    int numCores = thread::hardware_concurrency();
    for(int i = 0; i < numCores; i++)
    {
        _threads.push_back(new TaskThread(i, this));
    }
}

void TaskManager::submitTask(Task *task) {
    _tasks.add(task);
    _mutex.lock();
    _totalTasks++;
    _mutex.unlock();
}

void TaskManager::submitTasks(vector<Task*>* tasks) {
    for(Task* t : *tasks)
        _tasks.add(t);
    _mutex.lock();
    _totalTasks += tasks->size();
    _mutex.unlock();
}

void TaskManager::killThreads() {
    for(TaskThread* t : _threads)
        _tasks.add(new TerminateTask());
    for(TaskThread* t : _threads)
        t->join();
}

void TaskManager::waitForTasks() {
    unique_lock<mutex> lock(_mutex);
    while(_totalTasks > 0) {    //Wait for no more tasks being executed
        _cond.wait(lock);
    }
}

Task* TaskManager::takeTask() {
    return _tasks.take();
}

void TaskManager::notifyOfTaskCompletion() {
    _mutex.lock();
    _totalTasks--;
    _cond.notify_one();
    _mutex.unlock();
}